feat: add new 'modern' implementation of Fake Timers

SimenB opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

This is an alternative to #5171 and #7300. Difference from them is that in this PR Lolex is added as an opt-in alternative to the current implementation of fake timers, instead of replacing it.

I've added `jest.useFakeTimers('modern');` which will allow people to migrate test file by test file if they want the new system. People can also use `useFakeTimers('legacy')` if they want to keep using the old implementation.

The default if nothing is passed to `jest.useFakeTimers` (or just `fake` is passed in config) will be the legacy implementation in Jest 26, and we will swap the default in Jest 27. We have no plans to ever remove the legacy implementation.

Hopefully this means FB can migrate over over timer and we can remove Jest's implementation.

Fixes #3465
Fixes #5165
Closes #5171
Closes #7300

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test plan

Copied over the refactored unit test from #5171, and added an integration test.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

SimenB commented on e2e/timer-reset-mocks/after-reset-all-mocks/timerAndMock.test.js:
```diff
@@ -4,17 +4,17 @@ describe('timers', () => {
-    setImmediate(() => f());
-    jest.runAllImmediates();
-    expect(f.mock.calls.length).toBe(1);
+    setTimeout(f, 0);
```
non-related changes - I just kept them from the original PR since this works for both implementations 🙂

wmertens commented:
Just wanted to add that unlike Jest, Lolex implements proper mocking for the Node timers, returning an object with `.unref()`, so looking forward to this!

SimenB commented:
Jest also returns `ref` and `unref` if you run your test in the node environment. Has done since Jest 22: #4622

SimenB commented:
Rebased this after the great TS migration.

---

I'm in favour of landing this in 24 and flipping the default in 25. While it sucks to ship multiple implementations of the same functionality (fake timers), I think using Lolex is such an improvement that it's worth it. Then people will have a whole major to migrate to the new implementation, and they can do it file by file if they want.

Not sure if deprecation warnings are worth it (especially as this is per file, so there'd be a lot of them)

@jeysal @mattphillips @thymikee @rubennorte @scotthovestadt thoughts?

jeysal commented:
I haven't followed this in detail - is it likely that switching the default will break a lot of users or did it only happen at FB because it's used in a weird/specific/rare way?
If it's the former I think deprecation warnings would be good so people are aware of upcoming migration work and have time to already migrate to Lolex. If it's the latter I'd say they are not needed.

SimenB commented:
Summary of how this is breaking in #5171. I'd wager to say that if your test fails after moving to Lolex, it's because it depends on bugs/broken behavior in Jest (barring the case where we now mock more than before, but that's ok, I think)

SimenB commented:
Thanks for chiming in @quasicomputational!

I really like your suggestion and I think we should go with that. I'd call the current implementation `legacy` instead of `jest` though - sounds a bit more scary so hopefully people want to avoid it.

Related, I'd also like to _not_ call it `lolex` since the end goal is that it's an implementation detail. But calling it `modern` or something just seems like an unnecessary indirection

SimenB commented on packages/jest-fake-timers/src/FakeTimersLolex.ts:
```diff
@@ -0,0 +1,140 @@
+    this._maxLoops = maxLoops || 100000;
+
+    this._fakingTime = false;
+    this._lolex = lolexWithGlobal(global);
```
Wouldn't it work for you to just polyfill `queueMicrotask` in a setup file, then Lolex would pick it up and fake it?

---

We discussed this (passing it what to mock) briefly in the fall. IIRC the argument against allow you to pass in args was that you shouldn't be able to opt _out_ of certain methods - that would allow you to write tests in a way that would not reflect at all how it would run in the real world (beyond the fact you're faking time at all).

quasicomputational replied:
I think a dummy `queueMicrotask` in an environment would suffice, yeah. It'd feel a bit weird because you don't actually need to even attempt to polyfill it - just having that property on the object is enough! But if I let little things like that bother me, coding in JavaScript would drive me insane...

Being hard to use wrong is a good goal. And there's already an escape hatch in manually setting up Lolex in whatever crazy way you want it, so this isn't critical or anything.

quasicomputational commented:
Sounds good. Avoiding exposing the implementation detail in the name is an excellent principle. If I can bikeshed a bit, `pre-25` and `post-25`, or `v1` and `v2`, or something else explicitly versioned would be more future proof, just in case there's cause for another implementation switch down the line and suddenly `modern` isn't so modern.

vovkasm commented:
I'm highly interested in this. Does this feature has any ETA? Can I help?

SimenB commented:
I'll cry a thousand tears if I'm unable to land one of the 3 Lolex PRs for Jest 25, which is next up (no time estimate, though)

vovkasm commented:
Just to share with community (thought it was already mentioned in some of comments, but cant remember and without example), anyone can already use lolex as is. I use this setup for fully synchronous testing:

1. Setup file for all tests (allow to use fake microtasks queue in Promise implementation), setup.js:
```js
// Tune promises (we use Bluebird with direct scheduler so it can be mocked by lolex...)
global.Promise = require('bluebird')
global.Promise.setScheduler(function(fn) {
  setImmediate(fn) // or process.nextTick(fn)
})
```

2. Distinct file to allow import it early `testing/setupTestTimers.ts`:
```ts
import lolex from 'lolex'
const clock = lolex.install({ toFake: Object.keys(lolex.timers) as any })
export { clock }
```

3. Then in tests:
```ts
import 'jest'

import { clock } from 'testing/setupFakeTimers'

// code to test
function example() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(true)
    }, 1000)
  })
}

beforeEach(() => {
  clock.setSystemTime(Date.parse('2018-04-03T14:44:00Z')) // Can set mocked time for ex
})

afterEach(() => {
  clock.reset()
})

it('example resolves to true after one second', () => {
  const cb = jest.fn()
  example().then(cb)
  clock.tick(500)
  expect(cb).not.toHaveBeenCalled()
  clock.tick(501)
  expect(cb).toHaveBeenCalledWith(true)
})
```

Notes.
1. This will not (and cant actually in modern v8) mock native Promise and async/await, so you need polyfill custom (js-only) versions before your code will use them (I use bluebird in example and ban using of any async/await syntax in the project).
2. This will not work with `jest.fn().mockResolvedValue()` function because this function will be evaluated before mock applied and will use native Promise.

SimenB commented:
Current plan is to land this as default in Jest 25, but allow people to do `jest.useFakeTimers('legacy');` to opt-in to the old one. I'd rather now make the new implementation opt-in first, mostly so we don't have to juggle migration paths

SimenB commented:
Aight, rebased this and added the method from #8713. I'll change from a `lolex` string to a `legacy` string and flip the default, then land it when we start landing breaking changes for Jest 25. Hopefully we'll get a release out before it's been 2 years since #5171

SimenB commented:
Swapped defaults, pretty much ready to land this now. Just needs a :+1: from fb this won't break them horribly

shaodahong commented:
looking forward this merge

inad9300 commented:
Amazing work, @SimenB, thank you! Looking forward to this too.

JoshuaKGoldberg commented:
@SimenB are you crying a thousand tears?

More importantly: is there anything that can be done to speed this PR forward? I'm _really_ looking forward to lolex-style fake timers in Jest!

SQReder commented:
Bump for merge

SimenB commented:
My tears have all dried up 😞 

This will be in Jest 26 - it slipped from 25 since we wanted to make it as non-breaking as possible.

SimenB commented:
Correction, this will be in Jest 26, but as opt-in via `jest.useFakeTimers('modern')`, and we'll flip the default in 27 together with a codemod for those (fb) wanting to stay on the old implementation.

I'll land this this weekend or early next week

marcosvega91 commented:
Hi @SimenB with the new implementation there is a way to understand if the user is using fake or real time ? A function like `isFakeTime` ?
